### PR TITLE
Bound GTFS manager shutdown with a context

### LINKED ISSUE #736: fix: Add Timeout to Graceful Shutdown in GTFS Manager
## Description
`Manager.Shutdown()` in `internal/gtfs/gtfs_manager.go` calls `m.wg.Wait()` with no timeout, causing the server to hang indefinitely if any goroutine (a stuck real-time feed fetch) never returns.

## Current Behavior
```go
func (m *Manager) Shutdown() error {
    close(m.stopChan)
    m.wg.Wait() // blocks forever if any goroutine is stuck
    return nil
}
```

## Expected Behavior
`Shutdown()` should accept a `context.Context`. If the context deadline is
exceeded before all goroutines finish, it returns a descriptive error and
allows the caller to proceed rather than blocking forever.

## Proposed Fix
```go
func (m *Manager) Shutdown(ctx context.Context) error {
    close(m.stopChan)

    done := make(chan struct{})
    go func() {
        m.wg.Wait()
        close(done)
    }()

    select {
    case <-done:
        return nil
    case <-ctx.Done():
        return fmt.Errorf("shutdown timeout exceeded: %w", ctx.Err())
    }
}
```

The caller constructs the context with an
appropriate deadline before invoking shutdown:
```go
ctx, cancel := context.WithTimeout(context.Background(), 25*time.Second)
defer cancel()
if err := manager.Shutdown(ctx); err != nil {
    log.Warn("shutdown did not complete cleanly", "error", err)
}
```

## Acceptance Criteria
- [ ] `Shutdown` signature updated to `Shutdown(ctx context.Context) error`
- [ ] All call sites updated to pass a context with a timeout
- [ ] Unit test: shutdown completes normally when goroutines finish in time
- [ ] Unit test: shutdown returns an error when context deadline is exceeded
- [ ] `make lint` and `make test` both pass

### PR DESCRIPTION
**Pull request by omlahore**
Fixes #736

`Manager.Shutdown` called `wg.Wait()` with no deadline:

```go
func (manager *Manager) Shutdown() {
	manager.shutdownOnce.Do(func() {
		close(manager.shutdownChan)
		manager.wg.Wait() // no bound
		...
	})
}
```

A background worker that never returns, such as a real-time feed fetch stuck on a socket, blocks the whole shutdown path forever. The database never gets closed and the process never exits.

## What changed

`Shutdown` now takes a `context.Context`. It still waits for workers, but it stops waiting when the context is done, logs, and returns the wrapped context error. The database is closed either way, so a hung worker no longer costs a clean database close.

`cmd/api` gives it a 30 second budget and logs if shutdown does not finish inside that. The 21 test call sites pass `context.Background()`.

I kept the signature change the issue asked for rather than adding a second method, since `Shutdown` has one production caller and the rest are tests.

## Tests

`internal/gtfs/shutdown_timeout_test.go`:

- `TestShutdownReturnsWhenContextExpires` parks a worker that never returns, calls `Shutdown` with a 50ms context, and asserts it comes back wrapping `context.DeadlineExceeded`. It completes in 0.05s. Before this change the same test hangs until the go test timeout.
- `TestShutdownReturnsNilWhenWorkersFinish` covers the normal path
- `TestShutdownIsIdempotent` covers the second call, since `shutdownOnce` means only the first does the work

## Checks

- `go vet -tags "sqlite_fts5 sqlite_math_functions" ./...` clean
- `go vet -tags "purego" ./...` clean
- `go fmt ./...` clean
- `make test` green across all 14 packages

131 insertions, 29 deletions, and 22 of the changed lines are the mechanical call-site update.

## Unrelated, while I was in there

Issue #491 looks stale. `rate_limit_middleware.go` already does the float division and `math.Ceil` that it proposes, fixed in 28325848 on 2026-02-28. Might be worth closing.

## Summary by CodeRabbit

- **Bug Fixes**
  - GTFS shutdown now respects cancellation and deadline limits, preventing indefinite blocking.
  - The application logs incomplete shutdowns and ensures resources are closed.

- **Reliability**
  - Repeated shutdown requests are handled safely.
  - Shutdown errors are preserved and reported when completion is delayed or cannot finish within the allotted time.

- **Tests**
  - Added coverage for shutdown timeouts, successful completion, repeated shutdown calls, retained errors, and eventual resource cleanup.

### CODE REVIEW

**Comment by omlahore** [comment]
Rebased onto main. The Windows failure was not from this change: the branch predated 031bf7b6, so it was still running the old `search_stops` cross-agency fixture, which fails under the `purego` SQLite build that the Windows job uses. Reproduced it locally with `-tags=purego`, and `TestSearchStopsHandlerParentStationCrossAgencyReference` passes after the rebase. Full package is green under both tag sets.

File: internal/gtfs/gtfs_manager.go [hunk 1]
```diff
@@ -261,17 +261,39 @@ func (manager *Manager) SetGtfsURL(url string) {
...
 
 // Shutdown gracefully shuts down the manager and its background goroutines
-func (manager *Manager) Shutdown() {
+// Shutdown stops background workers and closes the database. It waits for the
+// workers to finish, but only until ctx is done: a real-time feed fetch that
+// never returns would otherwise block the caller forever. On timeout it logs,
+// returns the context error, and still closes the database.
+func (manager *Manager) Shutdown(ctx context.Context) error {
+	var err error
 	manager.shutdownOnce.Do(func() {
```
[the remark this answers is not in the document]
Comment (omlahore) [reply internal/gtfs/gtfs_manager.go:270]: Both correct, fixed in 686a7fa5.

The error was a per-call local while `shutdownOnce` runs the body once, so the first caller saw the timeout and everyone after it got `nil`. It is on the `Manager` now and returned from every call. `TestShutdownRepeatsTheFirstError` fails on the previous commit with `second Shutdown returned nil, want the first call's error`.

File: internal/gtfs/gtfs_manager.go [hunk 2]
```diff
@@ -261,17 +261,39 @@ func (manager *Manager) SetGtfsURL(url string) {
...
+		case <-ctx.Done():
+			err = fmt.Errorf("waiting for background workers: %w", ctx.Err())
+			logging.LogError(logger, "shutdown timed out, closing database anyway", err)
+		}
+
 		if manager.GtfsDB != nil {
-			if err := manager.GtfsDB.Close(); err != nil {
-				logger := slog.Default().With(slog.String("component", "gtfs_manager"))
-				logging.LogError(logger, "failed to close GTFS database", err)
+			if closeErr := manager.GtfsDB.Close(); closeErr != nil {
```
[the remark this answers is not in the document]
Comment (omlahore) [reply internal/gtfs/gtfs_manager.go]: Fixed in 686a7fa5, and this was the more embarrassing of the two since bounding shutdown is what the PR is for.

`GtfsDB.Close()` now races the context instead of running synchronously after the deadline has already passed, so a stuck query cannot push `Shutdown` past `ctx`.

I did not add the in-flight-query test. Holding a query open long enough to be deterministic needs a hook in the manager that exists only for the test, and the existing `TestShutdownReturnsWhenContextExpires` already pins the bound on the worker-wait path. Say if you would rather have it and I will add the seam.

**Comment by aaronbrethorst** [comment]
### Code review

Found 1 issue:

1. When the context times out, `Shutdown` closes the GTFS database while the workers it just gave up on are still running. The maintainer's review of the earlier attempt at this fix (#768) flagged exactly this as critical: "DB is closed while background goroutines may still be running ... On timeout, Shutdown returns the error but the DB stays open for the still-running goroutines". The suggested fix there was to close the DB inside the `wg.Wait()` goroutine, so it only closes after the workers exit. The code's other claim, "still closes the database", doesn't hold up either. On the timeout path `ctx.Done()` is already closed when the second `select` runs, and the `Close()` goroutine has only just started, so that `select` nearly always takes the `ctx.Done()` branch. `Close()` gets abandoned and "gave up waiting for the GTFS database to close" is logged. I tested this with a real `:memory:` `GtfsDB`, a parked worker, and a 5ms context: 200 of 200 runs logged "gave up waiting". A worker that queried after `Shutdown` returned got `sql: database is closed`. A realistic trigger is the 24h static reload, which runs under its own 5-minute `context.Background()` timeout and can outlast the 30s budget. All four new tests use `GtfsDB: nil`, so none of them runs this branch. Separately, the justifying comment says "sql.DB.Close waits for in-flight queries". `database/sql`'s `DB.Close` only closes idle connections and returns without waiting for connections in use.

https://github.com/OneBusAway/maglev/blob/6981c9604f83edb0ef51660dd9c9cd60f671f0c2/internal/gtfs/gtfs_manager.go#L263-L311

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by omlahore** [comment]
Hey, thanks. I guess you're right, it closed the DB with the workers still running. In 84fe790a the DB only closes from the goroutine that waits for the workers, and on timeout Shutdown still returns the ctx error. The new test uses a real :memory: db and fails every time on the old code 👍

**Review by aaronbrethorst — DISMISSED** [review]
Thanks for picking #736 back up. Bounding the wait is right, and keeping `shutdownErr` across calls, running the wait goroutine inside the `Once`, and leaving the HTTP-then-GTFS shutdown order unchanged are all good.

One blocker, and it's the same one I flagged as critical on #768. **On the timeout path, the database is closed while the workers we just gave up on are still running.**

- After `ctx.Done()` fires, `Shutdown` still starts `GtfsDB.Close()` in a goroutine.
- The second `select` then almost always takes `ctx.Done()` again, because the context is already done. So `Shutdown` logs "gave up waiting," but the close still goes through in the background, under any worker that's still running.
- That's realistic: the daily static reload runs under its own 5-minute context and can easily outlast a 30s budget. A worker that queries after that gets `sql: database is closed`, and a reload in the middle of writing gets its connection pulled out from under it.
- Separately, the comment's premise doesn't hold. `sql.DB.Close`'s doc says it waits for in-flight queries, but the implementation only closes idle connections and returns.

The fix is small:
- Close the database inside the goroutine that runs `manager.wg.Wait()`, so it only happens after the workers exit.
- On timeout, return the context error and leave the DB open. The process is exiting anyway.
- Add a test with a non-nil `GtfsDB` that parks a worker and asserts the DB is still usable after a timed-out `Shutdown`. All four current tests use `GtfsDB: nil`, so none of them run this branch.

Two optional cleanups while you're in there:
- The old `// Shutdown gracefully shuts down...` line is still above the new doc comment.
- HTTP shutdown and GTFS shutdown each get 30s, so up to 60s total. That's longer than Docker's default 10s stop grace period and Kubernetes' 30s.

Happy to re-review.

**Comment by omlahore** [comment]
Hey, I think this one crossed with 84fe790a too. The DB only closes from the goroutine that waits for the workers now, and the new test uses a real in-memory db with a parked worker. I left the 30s plus 30s budget alone, I wasn't sure if you wanted that changed here 🙏
